device/src/esp32_run.rs

```diff
@@ -411,36 +324,228 @@ where
+                        upstream_connection
+                            .set_state(UpstreamConnectionState::EstablishedAndCoordAck, &mut ui);
+                    }
+                    CoordinatorSendBody::Naming(naming) => match naming {
+                        frostsnap_comms::NameCommand::Preview(preview_name) => {
+                            ui.set_workflow(ui::Workflow::NamingDevice {
+                                old_name: name.clone(),
+                                new_name: preview_name.clone(),
+                            });
+                        }
+                        frostsnap_comms::NameCommand::Finish(new_name) => {
+                            ui.set_workflow(ui::Workflow::prompt(ui::Prompt::NewName {
+                                old_name: name.clone(),
+                                new_name: new_name.clone(),
+                            }));
+                        }
+                    },
+                    CoordinatorSendBody::Core(core_message) => {
+                        // FIXME: It is very inelegant to be inspecting
+                        // core messages to figure out when we're going
+                        // to be busy.
+                        match &core_message {
+                            CoordinatorToDeviceMessage::DoKeyGen { .. } => {
+                                ui.set_busy_task(ui::BusyTask::KeyGen)
+                            }
+                            CoordinatorToDeviceMessage::FinishKeyGen { .. } => {
+                                ui.set_busy_task(ui::BusyTask::VerifyingShare)
+                            }
+                            _ => { /* no workflow to trigger */ }
+                        }
+
+                        outbox.extend(
+                            signer
+                                .recv_coordinator_message(core_message.clone(), &mut rng)
+                                .expect("failed to process coordinator message"),
+                        );
                     }
+                    CoordinatorSendBody::Upgrade(upgrade_message) => match upgrade_message {
+                        CoordinatorUpgradeMessage::PrepareUpgrade {
+                            size,
+                            firmware_digest,
+                        } => {
+                            let upgrade_ = ota_config.start_upgrade(
+                                flash.flash_mut(),
+                                *size,
+                                *firmware_digest,
+                                active_firmware_digest,
+                            );
+
+                            upgrade = Some(upgrade_);
+                        }
+                        CoordinatorUpgradeMessage::EnterUpgradeMode => {}
+                    },
                 }
             }
 
-            if let Some(ui_event) = ui.poll() {
+            {
+                let staged_mutations = signer.staged_mutations();
```

Comment:
You accidentally duplicated a lot of code in this file

Reply:
Thanks for catching this!